sunkit_image/nafe.py

```diff
@@ -0,0 +1,411 @@
+            elif roi_f[0] <= 0 or roi_f[1] <= 0:
+                raise ValueError("Invalid roi_f given!", roi_f)
+            elif not isinstance(roi_p[0], int) or not isinstance(roi_p[1], int):
+                raise ValueError("roi_p should be a non-negative"
+                                 "integer numbers! roi_p given is:",
+                                 roi_p)
+            elif not isinstance(roi_f[0], int) or not isinstance(roi_f[1], int):
+                raise ValueError("roi_f should be a non-negative "
+                                 "integer numbers! roi_f given is:",
+                                 roi_f)
+            elif roi_p[0] + roi_f[0] > input_map.data.shape[0] or roi_p[1] + roi_f[1] > input_map.data.shape[1]:
+                print("Warning: part of frame is outside of given "
+                      "input_map! Clipping it.")
+                if roi_p[0] + roi_f[0] > input_map.data.shape[0]:
+                    roi_f[0] -= input_map.data.shape[0] - (roi_p[0] + roi_f[0])
+                if roi_p[1] + roi_f[1] > input_map.data.shape[1]:
+                    roi_f[1] -= input_map.data.shape[1] - (roi_p[1] + roi_f[1])
+                print("New frame size is:", roi_f)
+            map_data = input_map.data[roi_p[0]: roi_p[0] + roi_f[0], roi_p[1]: roi_p[1] + roi_f[1]]
+        else:
+            map_data = input_map.data
+
+        if nproc is None:
+            nproc = multiprocessing.cpu_count()
+        elif not isinstance(nproc, int) or nproc <= 0:
+            raise ValueError("nproc should be a non-negative integer number! "
+                             "nproc given is:", nproc)
+        elif nproc > multiprocessing.cpu_count():
+            print("Warning: nproc given is too big! Suggested value is:",
+                  multiprocessing.cpu_count())
+
+        if hbins is None:
+            # Freedman–Diaconis rule to calculate optimum number of
+            # historgam bins
+            iqr = np.subtract(*np.percentile(map_data, [75, 25]))
+            h = 2 * iqr / (n ** (1 / 3))
+            hbins = int(np.round((map_data.max() - map_data.min()) / h))
+        elif not isinstance(hbins, int) or hbins <= 0:
+            raise ValueError("hbins should be a non-negative integer number! "
+                             "hbins given is:", hbins)
+
+        if nrs < 0.0:
+            raise ValueError("nrs should be a non-negative real number! "
+                             "nrs given is:", nrs)
+
+        if gb < 0 or lb < 0 or gb + lb != 1:
+            raise ValueError("Invalid values for (gb, lb)!", (gb, lb))
+
+        if gw < 0 or lw < 0 or gw + lw != 1:
+            raise ValueError("Invalid values for (gw, lw)!", (gw, lw))
+
+        if not isinstance(n, int) or n % 2 == 0:
+            raise ValueError("n should be an odd integer. n given is:", n)
+
+        if len(c) <= 0:
+            raise ValueError("No constants(c) given! "
+                             "At least one should be given")
+
```

Comment:
This is a large amount of input checking, is this all necessary?

Reply:
Not really sure how it's done, some basic checks are required at least like n being positive integer and gamma being non-negative real.

Reply:
One of the reasons I went for this project was learning to develop high quality python code. My reference code-bases tended to do some input checking but I may have overdid it that's why I am not really sure.

Reply:
I'm not against the idea but it seems very overkill. We should at least error on values that make zero  sense. For others, we can warn.